src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -528,6 +572,668 @@ public override TResult[] ToArray()
+            }
+
+            public override IEnumerable<TResult> Where(Func<TResult, bool> predicate)
+            {
+                return new WhereEnumerableIterator<TResult>(this, predicate);
+            }
+
+            public override TResult[] ToArray()
+            {
+                var results = new TResult[_source.Count];
+                for (int i = 0; i < results.Length; i++)
+                {
+                    results[i] = _selector(_source[i], i);
+                }
+                return results;
+            }
+            
+            public int IndexOf(TResult item)
+            {
+                int index = 0;
+                EqualityComparer<TResult> cmp = EqualityComparer<TResult>.Default;
+                foreach (TSource sourceItem in _source)
+                {
+                    if (cmp.Equals(item, _selector(sourceItem, index))) return index;
+                    ++index;
+                }
+                return -1;
+            }
+            
+            public bool Contains(TResult item)
+            {
+                EqualityComparer<TResult> cmp = EqualityComparer<TResult>.Default;
+                int index = 0;
+                foreach (TSource sourceItem in _source)
+                {
+                    if (cmp.Equals(item, _selector(sourceItem, index))) return true;
+                    ++index;
+                }
+                return false;
+            }
+            
+            public TResult this[int index]
+            {
+                get { return _selector(_source[index], index); }
+                set { throw Error.NotSupported(); }
+            }
+            
+            public void CopyTo(TResult[] array, int arrayIndex)
+            {
+                ValidateCopyTo(array, arrayIndex, Count);
+                int index = 0;
+                foreach (TSource item in _source) array[arrayIndex++] = _selector(item, index++);
+            }
+            
+            public int Count
+            {
+                get { return _source.Count; }
+            }
+
+            public void Insert(int index, TResult item)
+            {
+                throw Error.NotSupported();
+            }
+    
+            public void RemoveAt(int index)
+            {
+                throw Error.NotSupported();
+            }
+    
+            public void Add(TResult item)
+            {
+                throw Error.NotSupported();
+            }
+    
+            public void Clear()
+            {
+                throw Error.NotSupported();
+            }
+    
+            public bool Remove(TResult item)
+            {
+                throw Error.NotSupported();
+            }
+    
+            public bool IsReadOnly
+            {
+                get { return true; }
+            }
+        }
+
+        internal sealed class SelectListIterator<TSource, TResult> : Iterator<TResult>, IList<TResult>
```

Comment:
This is an optimised class that could also be served by `SelectIListIterator<TSource, TResult>`. My instincts would have been not to include it, but as it replaces `WhereSelectListIterator<TSource, TResult>` it seemed perhaps best not to remove the degree to which that class optimises by using `List<TSource>.Enumerator` directly.